text/0000-extern-existential-type.md

```diff
@@ -0,0 +1,217 @@
+- Feature Name: extern-existential-type
+- Start Date: 2018-6-29
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+A version of #2071's `existential type` where the definition can live in a different crate than the declaration.
+This is a crucial tool untangling for untangling dependencies within `std` and other libraries at the root of the ecosystem concerning global resources.
+
+# Motivation
+[motivation]: #motivation
+
+We have a number of situations where one create defines an interface, and a different crate implements that interface with a canonical singleton:
+
+ - [`core::alloc::GlobalAlloc`](https://doc.rust-lang.org/nightly/core/alloc/trait.GlobalAlloc.html), chosen with [`#[global_allocator]`](https://doc.rust-lang.org/1.23.0/unstable-book/language-features/global-allocator.html)
+ - `panic_fmt` chosen with [`#[panic_implementation]`](https://github.com/rust-lang/rfcs/blob/master/text/2070-panic-implementation.md)
+ - The OOM hook, modified with [`std::alloc::{set,take}_alloc_error_hook`](https://doc.rust-lang.org/nightly/std/alloc/fn.set_alloc_error_hook.html)
+ - `std::hash:RandomState`, if https://github.com/rust-lang/rust/pull/51846 is merged, the `hashmap_random_keys` lang item.
+ - [`log::Log`](https://docs.rs/log/0.4.3/log/trait.Log.html) set with https://docs.rs/log/0.4.3/log/fn.set_logger.html
+
+Each of these is an instance of the same general pattern.
+But the solutions are all ad-hoc and distinct, adding burdening the user of Rust and rustc with needless extra information, and preventing more rapid prototyping.
+They also incur a run-time cost due to dynamism and indirection, which can lead to initializing bugs or bloat in space-constrained environments.
+
+`existential type` just covers the deferred definition of a type, and not the singleton value itself, but that is enough. For example, with global allocation:
+
+```rust
+// in `alloc`
+
+pub extern existential type Heap: Copy + Alloc + Default;
+
+struct Box<T, A: Alloc = Heap>;
+
+impl Box<T, A: Alloc> {
+    fn new_in(a: A) { .. }
+}
+
+impl Box<T, A: Alloc + Default = Heap> {
+    fn new() { Self::new_in(Default::default())    }
+}
+```
+
+```rust
+// in `jemalloc`
+
+struct JemallocHeap;
+
+impl Default for JemallocHeap {
+    fn default() { JemallocHeap }
+}
+
+impl Alloc for JemallocHeap {
+    fn alloc(stuff: Stuff) -> Thing {
+        ...
+    }
+}
+
+impl existential type alloc::Heap = JemallocHeap;
+```
+
+```rust
+// in crate making an rust-implemented local allocator global
+
+static GLOBALIZED_LOCAL_ALLOC = MyConcurrentLocalAlloc(..):
```

Comment:
Don't introduce `static` before definition of type; easier to read if type is introduced first.